mono/metadata/threads.c

```diff
@@ -4515,61 +4545,93 @@ mono_thread_execute_interruption (void)
 	mono_thread_info_clear_self_interrupt ();
 
 	/* If there's a pending exception and an AbortRequested, the pending exception takes precedence */
-	if (sys_thread->pending_exception) {
-		MonoException *exc;
-
-		exc = sys_thread->pending_exception;
-		sys_thread->pending_exception = NULL;
-
-		UNLOCK_THREAD (thread);
-		return exc;
-	} else if (thread->state & (ThreadState_AbortRequested)) {
-		UNLOCK_THREAD (thread);
-		g_assert (sys_thread->pending_exception == NULL);
-		if (thread->abort_exc == NULL) {
-			/* 
-			 * This might be racy, but it has to be called outside the lock
-			 * since it calls managed code.
-			 */
-			MONO_OBJECT_SETREF (thread, abort_exc, mono_get_exception_thread_abort ());
+	//pending_exception = sys_thread->pending_exception;
+	pending_exception = MONO_HANDLE_NEW_GET (MonoException, sys_thread, pending_exception);
+	if (!MONO_HANDLE_IS_NULL (pending_exception)) {
+		exc = pending_exception;
+		// sys_thread->pending_exception = NULL;
+		MONO_HANDLE_SETRAW (sys_thread, pending_exception, NULL);
+		goto exit;
+	} else if (thread->state & ThreadState_AbortRequested) {
+		g_assert (MONO_HANDLE_IS_NULL (pending_exception));
+
+		// Does the thread already have an abort exception?
+		// If not, create a new one and set it on demand.
+
+		// abort_exc = thread->abort_exc;
+		abort_exc = MONO_HANDLE_NEW_GET (MonoException, thread_handle, abort_exc);
```

Comment:
I think `abort_exc` scope could be just from here until the end of this `else if (...) { ... }` block.

Reply:
That seems reasonable. What I have is more analogous to the original, but the original didn't need/want to read it out into a local.